core/src/apps/bitcoin/sign_tx/bitcoin.py

```diff
@@ -197,21 +213,25 @@ async def step2_approve_outputs(self) -> None:
     async def step3_verify_inputs(self) -> None:
         # should come out the same as h_inputs_check, checked before continuing
         h_check = HashWriter(sha256())
+        expected_digest = None
 
         if self.taproot_only:
-            # All internal inputs are Taproot. We only need to verify external inputs. We can trust
-            # the amounts and scriptPubKeys, because if an invalid value is provided then all
-            # issued signatures will be invalid.
-            expected_digest = self.h_external_inputs
-            for i in range(self.tx_info.tx.inputs_count):
-                progress.advance()
-                if i in self.external:
-                    txi = await helpers.request_tx_input(self.tx_req, i, self.coin)
-                    writers.write_tx_input_check(h_check, txi)
-                    if not input_is_external_unverified(txi):
-                        # txi.script_pubkey checked in sanitize_tx_input
-                        assert txi.script_pubkey is not None
-                        await self.verify_external_input(i, txi, txi.script_pubkey)
+            # All internal inputs are Taproot. We only need to verify presigned external inputs.
+            # We can trust the amounts and scriptPubKeys, because if an invalid value is provided
+            # then all issued signatures will be invalid.
+            if self.has_presigned:
+                expected_digest = self.h_external_inputs
+                for i in range(self.tx_info.tx.inputs_count):
+                    if i in self.external:
+                        progress.advance()
+                        txi = await helpers.request_tx_input(self.tx_req, i, self.coin)
```

Comment:
> It seems to me that we can easily avoid streaming the inputs that have an ownership proof here. See [this draft](https://github.com/trezor/trezor-firmware/tree/onvej-sl/coinjoin-speed-step-3). Does it make sense?

The reason why I chose to use the `has_presigned` boolean value over a `set` of `presigned` indices is because it seemed like a premature optimization to use a `set`. It would optimize the signing speed only when all of the following conditions are satisfied:
- There is at least one presigned input.
- There are many external inputs with an ownership proof.
- All internal inputs are taproot.

So far we don't have a use-case for that. But it would be useful for CoinJoins where some participants don't have the ability to produce an ownership proof for their input and produce a presigned input. I am not against the optimization. I guess using a `set` vs. `bool` here won't make a relevant difference in terms of RAM.

Reply:
We can add this commit to the PR: https://github.com/trezor/trezor-firmware/commit/2700aa24fec578497ed1d993fb53654574254690

Reply:
@onvej-sl?

I added the commit to this PR as https://github.com/trezor/trezor-firmware/pull/2507/commits/0c24201a844b290657d3982b5d483edb41075105.